### PR TITLE
[Feature] Add Satrn

### PR DESCRIPTION
**Pull request by 2793145003**
Hi, 我在我自己的数据集（10w条数据）上使用了 SATRN（https://arxiv.org/pdf/1910.04396.pdf）的 small 版本，最开始的几天 acc 完全没有变化，但之后上升得很快，目前 word_acc=0.9245 并且还在缓慢上升中（执行时间 266:17:00）。
不过我没有足够的资源和时间使用公开数据集，也没有足够显存运行完整的模型，所以没有相应的指标，也不能保证 satrn_academic（论文中的结构）也能得到好的效果……

### CODE REVIEW

**Review by gaotongxiao — COMMENTED** [review]
Thanks for your contribution! Your work is excellent and really constructive to MMOCR!

I had compared your implementation with the official release and found some differences that should be addressed. Some are not significant but might slightly affect the performance. (We want to be conservative on this model due to its long training cycle) I had also commented on some parts that I'm not quite sure and we can discuss.

Meanwhile, we are also training the full Satrn model so we'll see if the model can match the claimed performance.

File: mmocr/models/textrecog/layers/transformer_layer.py [hunk 1]
```diff
@@ -188,6 +230,30 @@ def forward(self, x):
...
+    <https://arxiv.org/abs/1910.04396>`_
+    """
+
+    def __init__(self, d_in, d_hid, dropout=0.1):
+        super().__init__()
+        self.conv_1 = nn.Conv2d(d_in, d_hid, kernel_size=3, padding=1)
+        self.relu_1 = nn.ReLU(inplace=True)
+        self.dropout_1 = nn.Dropout(dropout)
+        self.conv_2 = nn.Conv2d(d_hid, d_in, kernel_size=3, padding=1)
+        self.dropout_2 = nn.Dropout(dropout)
```
[the remark this answers is not in the document]
Comment (2793145003) [reply mmocr/models/textrecog/layers/transformer_layer.py:245]: 他们的 DepthwiseConv2D 内容是 pass，我不知道该怎么写……
https://github.com/clovaai/SATRN/blob/master/src/networks/SATRN.py#L13-L16

[the remark this answers is not in the document]
Comment (gaotongxiao) [reply mmocr/models/textrecog/layers/transformer_layer.py:244]: It's inherited from `tf.keras.layers.DepthwiseConv2D`, which means `DepthwiseConv2D ` is just an alias for `tf.keras.layers.DepthwiseConv2D`

[the remark this answers is not in the document]
Comment (2793145003) [reply mmocr/models/textrecog/layers/transformer_layer.py]: 说得对，是我傻了……希望我改对了……

File: mmocr/models/textrecog/layers/transformer_layer.py [hunk 2]
```diff
@@ -217,6 +302,75 @@ def forward(self, x):
         return x + self.position_table[:, :x.size(1)].clone().detach()
 
 
+class Adaptive2DPositionalEncoding(nn.Module):
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/layers/transformer_layer.py:305]:
```suggestion
class Adaptive2DPositionalEncoding(BaseModule):
```

File: mmocr/models/textrecog/layers/transformer_layer.py [hunk 3]
```diff
@@ -217,6 +302,75 @@ def forward(self, x):
...
+      Licensed under the Apache License, Version 2.0 (the "License");
+    Args:
+        d_hid (int): Dimensions of hidden layer.
+        n_height (int): Max height of the 2D feature output.
+        n_width (int): Max width of the 2D feature output.
+        dropout (int): Size of hidden layers of the model.
+    """
+
+    def __init__(self, d_hid=512, n_height=100, n_width=100, dropout=0.1):
+        super().__init__()
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/layers/transformer_layer.py:318]:
```suggestion
    def __init__(self,
                 d_hid=512,
                 n_height=100,
                 n_width=100,
                 dropout=0.1,
                 init_cfg=[dict(type='Xavier', layer='Conv2d')]):
        super().__init__(init_cfg=init_cfg)
```

**Review by gaotongxiao — COMMENTED** [review]
Changed the config to get a closer implementation to the official implementation.

**Comment by gaotongxiao** [comment]
For now, I've trained your original Satrn implementation on Synthtext + Mjsynth for 5 epochs, but only got ~ 10%-20% accuracy on test sets. The modified version is still in its first training epoch but gets an average loss of ~0.02 which is 10 times smaller than the original one. I think we are on the right track.

**Comment by gaotongxiao** [comment]
Well, here's the best result after the 4th epoch:

|IIIT5k|SVT|IC13|IC15|SVTP|CT80|
|--------|-----|------|-------|--------|-------|
|0.9607|0.9351|0.9567|0.8406|0.8853|0.9028|

It's even better than what the original paper have claimed. I feel the model is good enough for now, and I will continue to wrap it up and publish this model soon. Good job!

**Comment by 2793145003** [comment]
好耶！感谢大佬的review和指点！

File: mmocr/models/textrecog/backbones/shallow_cnn.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import torch.nn as nn
+from mmcv.runner import BaseModule
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/shallow_cnn.py:2]:
```suggestion
from mmcv.runner import BaseModule
from mmcv.cnn import ConvModule
```

**Review by gaotongxiao — COMMENTED** [review]
Final code style pass, just a bit of refactoring. I'll push a new readme and `ocr.py` to this PR soon.
BTW, Github now supports applying multiple suggestions in a single commit :)

**Review by gaotongxiao — APPROVED** [review]

**Comment by gaotongxiao** [comment]
@innerlee @cuhk-hbsun Any general comments? If not, it will be merged tomorrow

**Comment by gaotongxiao** [comment]
您好@2793145003 ，请问您有兴趣加入 MMOCR 的开发者群吗？如果有的话，可以在 OpenMMLab 的官方 QQ 群里 (1447762544) 联系群主 OpenMMLab 小助手，他会把您拉进开发者群。期待您的加入！

**Comment by 2793145003** [comment]
> 您好@2793145003 ，请问您有兴趣加入 MMOCR 的开发者群吗？如果有的话，可以在 OpenMMLab 的官方 QQ 群里 (1447762544) 联系群主 OpenMMLab 小助手，他会把您拉进开发者群。期待您的加入！

没有搜到这个群……我ID就是QQ号，可以直接拉我：）
